### PR TITLE
[PoC] Simplify state management by not using Redux

### PR DESCRIPTION
**Pull request by andreypopp**
Using Redux is definitely an overhead for such a simple state model.

I propose another approach which uses [derivable](https://github.com/ds300/derivablejs) library for managing an app state.

### CODE REVIEW

**Comment by Termina1** [comment]
I am definitely agree, since this project was created by to illustrate some redux features, I needed it, now we can use something much simpler.

The only features I care about: to have as less deps as possible and vanilla webpack hot-reload. So, I would be happy to merge it if it's ready.

**Comment by andreypopp** [comment]
Yeah, hmr works for presentation contents (that's what I've tested). Obviously I removed `replaceReducer` HMR hook so no way to change state applying logic on the fly.

Btw. history doesn't seem to work (in original implementation too).

**Comment by Termina1** [comment]
Can we preserve state between reloads?

**Comment by andreypopp** [comment]
Yes, because it's external to components.

**Comment by andreypopp** [comment]
Let's not merge this before history behaviour is fixed.

File: js/AppState.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+export const state = Derivable.struct({mode, slide});
+
+/**
+ * Change currently active slide.
+ */
+export function changeActiveSlide(nextSlide) {
+  const hash = '#' + nextSlide;
+  // XXX: Because browser bugs with :target we must do this instead of fancy
+  // History.location.swap(...) :(
+  window.location.hash = hash;
```
Comment (andreypopp) [inline js/AppState.js:31]: What a bummer :(

File: js/History.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+history.listen(nextLocation => {
+  if (nextLocation.action === 'POP') {
+    location.set(nextLocation);
+  }
+});
+
+// Each time someone updates location atom we push location to history.
+location.react(nextLocation => {
+  history.push(nextLocation);
+}, {skipFirst: true});
```
Comment (andreypopp) [inline js/History.js:24]: This entire module could be a separate useful tiny npm package.

File: js/Persistence.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  if (currentValue === undefined) {
+    currentValue = initialValue;
+  }
+  const value = Derivable.atom(currentValue);
+  // Each time someone writes into atom we write value into localStorage.
+  value.react(val => {
+    write(key, val);
+  }, {skipFirst: true});
+  return value;
+}
```
Comment (andreypopp) [inline js/Persistence.js:38]: This entire module could be a separate useful tiny npm package.

File: js/components/Slide.react.js [hunk 4]
```diff
@@ -1,18 +1,19 @@
...
 export default class Slide extends Component {
 
   onClick() {
-    const dispatch = this.props.dispatch;
-    dispatch(changeActiveSlide(this.props.index));
-    dispatch(changeMode('full'));
+    Derivable.transact(() => {
+      AppState.changeActiveSlide(this.props.index);
+      AppState.changeMode('full');
+    });
```
Comment (andreypopp) [inline js/components/Slide.react.js:11]: It's really a bad idea to dispatch multiple actions at once. Should refactor into `zoomSlide` action or something like this.

File: js/components/Slide.react.js [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
   }
 
   render() {
-    var cls = '';
-    if (this.props.index == this.props.data.slide) {
+    let cls = '';
```
Comment (andreypopp) [inline js/components/Slide.react.js:15]: Sorry for unrelated change but linter was complaining really loud.

File: makewebpackconfig.js [hunk 6]
```diff
@@ -57,6 +57,7 @@ module.exports = function(options) {
       path.resolve(__dirname, 'js/app.js') // Start with js/app.js...
     ];
     cssLoaders = 'style-loader!css-loader!postcss-loader';
+    devtool = 'cheap-module-eval-source-map';
```
Comment (andreypopp) [inline makewebpackconfig.js:60]: Again, sorry for unrelated change but it's so useful in dev.

File: js/Persistence.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ */
+export function atom(key, initialValue) {
+  let currentValue = read(key);
+  if (currentValue === undefined) {
+    currentValue = initialValue;
+  }
+  const value = Derivable.atom(currentValue);
+  // Each time someone writes into atom we write value into localStorage.
+  value.react(val => {
+    write(key, val);
```
Comment (andreypopp) [inline js/Persistence.js:35]: We probably need to do `setTimeout` here so we won't block update.

File: js/AppState.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ * We store current mode in persistent atom (syncs with localStorage).
+ */
+const mode = Persistence.atom('__react_shower_mode__', 'list');
+
+/**
+ * App state is a {slide, mode}
+ *
+ * We expose struct derivation to prevent direct modification of the state.
+ */
+export const state = Derivable.struct({mode, slide});
```
Comment (andreypopp) [inline js/AppState.js:22]: I really love this: super idiomatic. We define app state as a composition of different pieces.

File: js/components/Deck.react.js [hunk 9]
```diff
@@ -65,21 +74,19 @@ class Deck extends Component {
...
+    var {slide, mode} = AppState.state.get();
     var children = React.Children.map(this.props.children, (child) =>
       React.cloneElement(child, {
         index: index++,
-        data: {
-          slide: this.props.slide,
-          mode: this.props.mode
-        },
-        dispatch: this.props.dispatch
+        data: {slide, mode},
```
Comment (Termina1) [inline js/components/Deck.react.js]: We should not forget to delete this.

File: js/components/Deck.react.js [hunk 10]
```diff
@@ -65,21 +74,19 @@ class Deck extends Component {
...
       })
     );
 
-    var progress = (this.props.slide - 1)/(getCount(this.props.children) - 1) * 100;
+    var progress = (slide - 1)/(getCount(this.props.children) - 1) * 100;
+    console.log(slide, mode);
```
Comment (Termina1) [inline js/components/Deck.react.js:86]: We should not forget to delete this.

**Comment by Termina1** [comment]
@andreypopp actually, it seems that history is working now? Also, this fixes #7. Can we merge it?

**Comment by andreypopp** [comment]
Yeah, I got history working. Sorry for not mentioning that.

**Pull request merged**